consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -0,0 +1,160 @@
+// Package cruisectl implements a "cruise control" system for Flow by adjusting
+// nodes' block rate delay in response to changes in the measured block rate.
+//
+// It uses a PID controller with the block rate as the process variable and
+// the set-point computed using the current view and epoch length config.
+package cruisectl
+
+import (
+	"time"
+
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// measurement represents one measurement of block rate and error.
+// A measurement is taken each time the view changes for any reason.
+// Each measurement measures the instantaneous and exponentially weighted
+// moving average (EWMA) block rates, computes the target block rate,
+// and computes the error terms.
+type measurement struct {
+	view            uint64    // v       - the current view
+	time            time.Time // t[v]    - when we entered view v
+	blockRate       float64   // r[v]    - measured instantaneous block rate at view v
+	aveBlockRate    float64   // r_N[v]  - EWMA block rate over past views [v-N, v]
+	targetBlockRate float64   // r_SP[v] - computed target block rate at view v
+	proportionalErr float64   // e_N[v]  - proportional error at view v
+	integralErr     float64   // E_N[v]  - integral of error at view v
+	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v
+}
+
+// epochInfo stores data about the current and next epoch. It is updated when we enter
+// the first view of a new epoch, or the EpochSetup phase of the current epoch.
+type epochInfo struct {
+	curEpochFinalView        uint64
+	curEpochTargetSwitchover time.Time
+	nextEpochFinalView       *uint64
+	epochFallbackTriggered   *atomic.Bool
+}
+
+// BlockRateController dynamically adjusts the block rate delay of this node,
+// based on the measured block rate of the consensus committee as a whole, in
+// order to achieve a target overall block rate.
+type BlockRateController struct {
+	component.Component
+
+	config *Config
+	state  protocol.State
+	log    zerolog.Logger
+
+	lastMeasurement *measurement    // the most recently taken measurement
+	blockRateDelay  *atomic.Float64 // the block rate delay value to use when proposing a block
+	epochInfo
+
+	viewChanges chan uint64       // OnViewChange events           (view entered)
+	epochSetups chan *flow.Header // EpochSetupPhaseStarted events (block header within setup phase)
+}
+
+// NewBlockRateController returns a new BlockRateController.
+func NewBlockRateController(log zerolog.Logger, config *Config, state protocol.State) (*BlockRateController, error) {
+
+	ctl := &BlockRateController{
+		config:      config,
+		log:         log,
+		state:       state,
+		viewChanges: make(chan uint64, 1),
+		epochSetups: make(chan *flow.Header, 1),
```

Comment:
I would be more comfortable, if the channel capacity was a bit higher specifically for the `viewChanges` channel. I think it doesn't harm to be able to buffer a few more events if the node is running at full capacity (e.g. when catching up)
* I would probably increase `epochSetups` to 3 and
* `viewChanges` to something like 10

I think the latter is specifically relevant because channels are _not_ atomically consistent, i.e. we might drop event during normal operations if capacity is too low.

Reply:
https://github.com/onflow/flow-go/pull/4332/commits/6b6d78cb6ee4fc4575903453dbcc972f1c0e4584